parlai/core/worlds.py

```diff
@@ -975,6 +975,202 @@ def shutdown(self):
         self.world.shutdown()
 
 
+class DynamicBatchWorld(World):
+    def __init__(self, opt: Opt, world: Union[DialogPartnerWorld, MultiWorld]):
+        super().__init__(opt)
+        self.opt = opt
+
+        # size of the buffer we will use to find worlds
+        self._BUFFER_SIZE = 1024
+
+        # agents is a placeholder just for super.reset
+        self.agents = []
+
+        # check some assumptions
+        if isinstance(
+            world, (ExecutableWorld, BatchWorld, HogwildWorld, MultiAgentDialogWorld)
+        ):
+            raise TypeError(
+                'World must be a DialogPartnerWorld or a '
+                'MultiWorld of DialogPartnerWorld'
+            )
+
+        if not hasattr(world, 'agents'):
+            raise TypeError("World doesn't have access to the agents.")
+
+        if len(world.agents) != 2:
+            raise AssertionError(
+                "Dynamic batch only works in a fixed dialog world with two agents."
+            )
+
+        if not hasattr(world.agents[1], 'batch_act'):
+            raise TypeError("Model agent doesn't have batch_act.")
+
+        self.truncate = opt.get('text_truncate', None) or opt.get('truncate', None)
+        if self.truncate is None or self.truncate < 0:
+            raise ValueError(
+                'You must use --text-truncate or --truncate in order to use '
+                'dynamic batching.'
+            )
+
+        # TODO: check to ensure the agent has self_observe
+        shared = world.share()
+        self.world = world
+        # TODO: maybe generalize this
+        self.max_size = self.truncate * opt['batchsize']
+
+        # buffer worlds
+        self.worlds = [
+            shared['world_class'](opt, None, shared) for _ in range(self._BUFFER_SIZE)
+        ]
+
+        self.reset()
+
+    def reset(self):
+        super().reset()
+        self._obs = [None for _ in range(self._BUFFER_SIZE)]
+        self._scores = [None for _ in range(self._BUFFER_SIZE)]
+
+        self.number_parleys = 0
+        self.total_exs = 0
+        self.world.reset()
+        for w in self.worlds:
+            w.reset()
+
+    def reset_metrics(self):
+        super().reset_metrics()
+        self.world.reset_metrics()
+        for w in self.worlds:
+            w.reset_metrics()
+
+    def epoch_done(self):
+        return all(w.epoch_done() for w in self.worlds)
+
+    def num_examples(self):
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        return self.world.num_episodes()
+
+    def _ceil(self, n, r=4):
+        """
+        Round to the nearest multiple of 4.
```

Comment:
why make `r` an argument if we only ever use it for 4?

Reply:
Because I used think it was 8. :D